ext/standard/string.c

```diff
@@ -3245,22 +3245,25 @@ PHPAPI zend_string *php_str_to_str(const char *haystack, size_t length, const ch
    Translates characters in str using given translation tables */
 PHP_FUNCTION(strtr)
 {
-	zval *from;
-	zend_string *str;
+	zend_string *str, *from_str = NULL;
+	HashTable *from_ht = NULL;
 	char *to = NULL;
 	size_t to_len = 0;
 	int ac = ZEND_NUM_ARGS();
 
 	ZEND_PARSE_PARAMETERS_START(2, 3)
 		Z_PARAM_STR(str)
-		Z_PARAM_ZVAL(from)
+		Z_PARAM_STR_OR_ARRAY_HT(from_str, from_ht)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_STRING(to, to_len)
 	ZEND_PARSE_PARAMETERS_END();
 
-	if (ac == 2 && Z_TYPE_P(from) != IS_ARRAY) {
+	if (ac == 2 && from_ht == NULL) {
 		zend_type_error("The second argument is not an array");
 		return;
+	} else if (ac != 2 && from_str == NULL) {
+		zend_type_error("The second argument is not a string");
+		return;
```

Comment:
Possibly we can improve these messages along these lines?
"If two arguments are passed, the second argument must be an array"
"If three arguments are passed, the second argument must be a string"

Reply:
Yes! I also didn't like them, but couldn't come up with a message good enough. :)

Reply:
Fixed. I only have to figure it out why the windows test fail.